homeassistant/components/tibber/sensor.py

```diff
@@ -410,86 +406,87 @@ def __init__(
-            )
-        )
-
     @property
     def available(self):
         """Return True if entity is available."""
         return self._tibber_home.rt_subscription_running
 
     @callback
-    def _set_state(self, state, timestamp):
-        """Set sensor state."""
-        if (
-            state < self._attr_state
-            and self.entity_description.reset_type == ResetType.DAILY
-        ):
-            self._attr_last_reset = dt_util.as_utc(
-                timestamp.replace(hour=0, minute=0, second=0, microsecond=0)
-            )
-        if (
-            state < self._attr_state
-            and self.entity_description.reset_type == ResetType.HOURLY
-        ):
-            self._attr_last_reset = dt_util.as_utc(
-                timestamp.replace(minute=0, second=0, microsecond=0)
-            )
+    def _handle_coordinator_update(self) -> None:
+        if not (live_measurement := self.coordinator.get_live_measurement()):  # type: ignore
+            return
+        state = live_measurement.get(self.entity_description.key)
+        if state is None:
+            return
+        timestamp = dt_util.parse_datetime(live_measurement["timestamp"])
+        if timestamp is not None and state < self._attr_state:
+            if self.entity_description.reset_type == ResetType.DAILY:
+                self._attr_last_reset = dt_util.as_utc(
+                    timestamp.replace(hour=0, minute=0, second=0, microsecond=0)
+                )
+            elif self.entity_description.reset_type == ResetType.HOURLY:
+                self._attr_last_reset = dt_util.as_utc(
+                    timestamp.replace(minute=0, second=0, microsecond=0)
+                )
+        if self.entity_description.key == "powerFactor":
+            state *= 100.0
         self._attr_state = state
         self.async_write_ha_state()
 
 
-class TibberRtDataHandler:
+class TibberRtDataCoordinator(update_coordinator.DataUpdateCoordinator):
     """Handle Tibber realtime data."""
 
     def __init__(self, async_add_entities, tibber_home, hass):
         """Initialize the data handler."""
         self._async_add_entities = async_add_entities
         self._tibber_home = tibber_home
         self.hass = hass
-        self._entities = {}
+        self._all_sensors = list(RT_SENSORS)
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=tibber_home.info["viewer"]["home"]["address"].get(
+                "address1", "Tibber"
+            ),
+        )
 
-    async def async_callback(self, payload):
-        """Handle received data."""
-        errors = payload.get("errors")
-        if errors:
-            _LOGGER.error(errors[0])
-            return
-        data = payload.get("data")
-        if data is None:
-            return
-        live_measurement = data.get("liveMeasurement")
-        if live_measurement is None:
+        self._async_remove_device_updates_handler = self.async_add_listener(
```

Comment:
Adding the callback that creates the entities as a listener here means we're going to poll the api even though the user has disabled all added entities. One of the benefits of the coordinator is that it will turn of polling if there are no enabled entities (listeners).

Reply:
There is no polling. It is a websocket where we receive data.
This will add new sensors, but not update any existing one.

Reply:
Any better way to solve it?

Reply:
Ok. I don't see the immediate benefit of the coordinator here. What is the reason we're adding it?

Reply:
This comment: https://github.com/home-assistant/core/pull/53357#issuecomment-885205065
And my follow up question on discord :)
![image](https://user-images.githubusercontent.com/650502/127373598-8e8cf490-4245-49d0-b8aa-cdbec8fdd7d4.png)

Reply:
Ok. I still don't see where the benefit lies in the PR implementation. The only feature of the coordinator this PR is using is the listener registration and manual callback.

Reply:
@balloob Any comment here?
I think Martin has a valid point, but I might have misunderstood something.

Reply:
So this is what I wrote when I suggested it:

> Right now the realtime data handler processes all the info instead of leaving that up to the entities. This means that even if an entity is disabled, the data is still processed and signals sent.

I think that it's still an improvement, albeit a minor one: the data source and the entity are directly connected instead of using a dispatcher.

Reply:
Ok, let's go ahead.